packages/client/src/sync/fetcher/trienodefetcher.ts

```diff
@@ -0,0 +1,578 @@
+  const buf = new Uint8Array(arr.length / 2)
+  for (let i = 0; i < buf.length; i++) {
+    let q = i * 2
+    buf[i] = (arr[q] << 4) + arr[++q]
+  }
+
+  return buf
+}
+
+/**
+ * Converts each nibble into a single byte
+ *
+ * @param arr Nibble typed nibble array
+ * @returns Uint8Array typed byte array
+ */
+function nibbleTypeToByteType(arr: Nibbles): Uint8Array {
+  const l = arr.length
+  const buf = new Uint8Array(l)
+  for (let i = 0; i < buf.length; i++) {
+    buf[i] = arr[i]
+  }
+
+  return buf
+}
+
+/**
+ * Turns each byte into a single nibble, only extracting the lower nibble of each byte
+ *
+ * @param key Uint8Array typed byte array
+ * @returns Nibble typed nibble array
+ */
+const byteTypeToNibbleType = (key: Uint8Array): Nibbles => {
+  const bkey = toBytes(key)
+  const nibbles = [] as Nibbles
+
+  for (let i = 0; i < bkey.length; i++) {
+    const q = i
+    nibbles[q] = bkey[i] % 16
+  }
+
+  return nibbles
+}
+
+/**
+ * Takes a string path and extends it by the given extension nibbles
+ *
+ * @param path String node path
+ * @param extension nibbles to extend by
+ * @param retType string indicating whether to return the key in "keybyte" or "hex" encoding
+ * @returns hex-encoded key
+ */
+const pathToHexKey = (path: string, extension: Nibbles, retType: string): Uint8Array => {
+  const b = hexToBytes(path)
+  const n = byteTypeToNibbleType(b)
+  if (retType === 'hex') {
+    return nibbleTypeToByteType(n.concat(extension))
+  } else if (retType === 'keybyte') {
+    return nibbleTypeToPackedBytes(n.concat(extension))
+  }
+  throw Error('retType must be either "keybyte" or "hex"')
+}
+
+type TrieNodesResponse = Uint8Array[] & { completed?: boolean }
+
+/**
+ * Implements an snap1 based bytecode fetcher
+ * @memberof module:sync/fetcher
+ */
+export interface TrieNodeFetcherOptions extends FetcherOptions {
+  root: Uint8Array
+  accountTrie?: Trie
+  codeTrie?: Trie
+  accountToStorageTrie?: Map<String, Trie>
+
+  /** Destroy fetcher once all tasks are done */
+  destroyWhenDone?: boolean
+}
+
+export type JobTask = {
+  pathStrings: string[] // paths kept in keybyte encoding until they are ready to send out
+  paths: Uint8Array[][] // paths to nodes for requesting from SNAP API kept in compact encoding
+}
+
+type FetchedNodeData = {
+  parentHash: string
+  deps: number
+  nodeData: Uint8Array
+  path: string
+  // storageNodes?: Uint8Array[] // if fetched node is account node with a storage component, storage node data is stored here until it is ready to be put
+  pathToStorageNode?: Map<string, Uint8Array>
+}
+
+type NodeRequestData = {
+  nodeHash: string
+  nodeParentHash: string
+  parentAccountHash?: string // for leaf account nodes that contain a storage component
+}
+
+export class TrieNodeFetcher extends Fetcher<JobTask, Uint8Array[], Uint8Array> {
+  protected debug: Debugger
+  root: Uint8Array
+  // Holds all paths and nodes that need to be requested
+  pathToNodeRequestData: OrderedMap<string, NodeRequestData>
```

Comment:
for storage path here is the path in its storage trie?

Reply:
A path is represented as a string of 0 to 129 characters. The first 64 are used to represent the path in the account trie. The final 64 are used to represent the path in the storage trie. A forward slash ('/') is used as a separator. This format is referred to as a "sync" or "stacked" path, representing the full path to a node in an account or storage trie.